codetransformer/decompiler.py

```diff
@@ -1291,26 +1352,57 @@ def make_function(function_builders, *, closure):
 
     return ast.FunctionDef(
         name=name.split('.')[-1],
-        args=ast.arguments(
-            args=[ast.arg(arg=a, annotation=annotations.get(a)) for a in args],
-            kwonlyargs=[
-                ast.arg(arg=a, annotation=annotations.get(a)) for a in kwonly
-            ],
-            defaults=defaults,
-            kw_defaults=list(map(kw_defaults.get, kwonly)),
-            vararg=None if varargs is None else ast.arg(
-                arg=varargs, annotation=annotations.get(varargs),
-            ),
-            kwarg=None if varkwargs is None else ast.arg(
-                arg=varkwargs, annotation=annotations.get(varkwargs)
-            ),
+        args=make_function_arguments(
+            args,
+            kwonly,
+            varargs,
+            varkwargs,
+            defaults,
+            kw_defaults,
+            annotations,
         ),
-        body=pycode_to_body(co, DecompilationContext(in_function=True)),
+        body=pycode_to_body(co, DecompilationContext(in_function_block=True)),
         decorator_list=decorators,
         returns=annotations.get('return'),
     )
 
 
+def make_function_arguments(args,
+                            kwonly,
+                            varargs,
+                            varkwargs,
+                            defaults,
+                            kw_defaults,
+                            annotations):
+    """
+    Make an ast.arguments from the args parsed out of a code object.
+    """
+    return ast.arguments(
+        args=[ast.arg(arg=a, annotation=annotations.get(a)) for a in args],
+        kwonlyargs=[
+            ast.arg(arg=a, annotation=annotations.get(a)) for a in kwonly
+        ],
+        defaults=defaults,
+        kw_defaults=list(map(kw_defaults.get, kwonly)),
+        vararg=None if varargs is None else ast.arg(
+            arg=varargs, annotation=annotations.get(varargs),
+        ),
+        kwarg=None if varkwargs is None else ast.arg(
+            arg=varkwargs, annotation=annotations.get(varkwargs)
+        ),
+    )
+
+
+def make_closure_cells(stack_builders):
+    cells = make_expr(stack_builders)
+    if not isinstance(cells, ast.Tuple):
+        raise DecompilationError(
+            "Expected an ast.Tuple of closure cells, "
+            "but got %s" % cells,
```

Comment:
@Vgr255 `make_expr` should always return an ast expression object.  This is intended to guard against manually-crafted or otherwise malformed bytecode that produces a non-tuple expression.

If we wanted to guard against bugs in `make_expr`, then we could make your suggested change, or we could explicitly format in a length-1 tuple, but I think the "correct" behavior here is actually to crash.

In general, `DecompilationError` should be used to signal that we can't handle the provided bytecode, not that there are bugs in our code.